✨Declare a GraphGen using GraphQL schema definition language (SDL)

cowboyd opened the pull request:
## Motivation

Graphgen has proven itself at this point for being an extraordinarily useful tool for simulating data across multiple services.  We have used it to seed a unified fixture that can drive a federation of integrated services including LDAP, Auth0, GraphQL, and REST-based micro services. However, as it stands, the graph generator itself must be expressed in an ad-hoc, complex, and barely coherent low-level format that resembled a graph database raw record structure rather than a coherent data model.  Specifically:

### Simple data distributions are awkward to implement

Because any data distribution can be attached to any relationship, and because they have different conventions depending on if it is one or many, picking the right distribution is confusing. This is because you have to match the type of distribution with the type of the relationship. a single relationship (Person.githubAccount) is a probablility distribution whose output is _presence_. A person has a probability of having a github account. It is either true, or false. On the other hand, a many relationship (Person.friends) is probability distribution over the _size_ of ther relationship, so an integer between zero and infinity. Which do you pick? `normal()`, `weighted()`, `uniform()` `constant()`?

### Passing relationship presets is difficult to get right.

When you want to create an object with a preset value so that your tests are co-located with your data, you pass in a record with that data. For simple scalar properties, this is straighforward:

``` js
createVertex(graph, "Person", {
  name: "Bob"
});
```

However, for related vertices, it is not so simple. You have to actually name the edge that you are traversing in order for graphgen to resolve which preset you are talking about.

``` javascript
createVertex(graph, "Person", {
  name: "Bob",
  "Person.manager": {
    name: "Judith"
  }
});
```

which is based on the assumption that there is an edge type call `"Person.manager"` from `Person` to `Person`, but this confuses the graph layer with they way that the graph will be consumed at runtime.

### Conventions for constructing inverse relationships have about 4 moving parts

To make the above example work, you need at a minimum the following graph definition:

``` javascript
createGraph({
  edge: [{
    name: Person.manager',
    from: 'Person',
    to: 'Person'
  }],
  vertex: [{
    name: 'Person',
    relationships: [{
      type: 'Person.manager',
      direction: 'from',
      size: constant(1),
    }]
  }],
});
```

You have to manually create the edge type first. Then you have to define the relationship and the distribution, and also include the direction.

###  Working with generated data is extremely difficult and involves manually de-referencing related nodes from the `graph` object.

At some point you have to consume your graph. But doing this is extremely difficult. For example:

``` javascript
let person: Vertex = createVertex("Person");
let edges = graph.from[person.id]
assert(edges);
let [managerEdge] = edges.filter(edge => edge.type === "Manager");
assert(managerEdge);
let manager = graph.vertices[manageredge.to];
assert(manager);
```

Fine if you're meta-programing, but as something to do on the daily? 🤮🤮🤮

### Different data generation paths depending on which direction you are traversing an edge from.

This is particularly not-great, but in order to allow customization of data based on other data, graphgen supports using _different_ data distributions based on which direction a relationship is being generated from. This was a bad solution for a very hard problem to begin with, and should, in my opinion, not be used anymore.

### Typing is non-existent

What you get back from `createVertex()` is `unknown` and `any`. That's not particularly helpful both in terms of the shape of the primitive data, but also, it's no help when it comes to dereferencing.


### Conclusion

Extending and maintaining the grapghen for a project is loaded with pitfalls and footguns and it places an unreasonable burden on developers to understand the inner complexities of the graph format. Ideally, developers should be able to quickly express the shape of their simulated in a single place and then start pumping out networks of data.

Most of these problems stem from the problem that while we have a graph format, and declarations of how it should be generated as you traverse it, there is no schema indicating how this data will be consumed. The best illustration of this is the dissonance between how you dereference data depending on which context you are using it from. For example, if I have two types `Person` and `Account` conceptually a Person _has many_ accounts. And if I'm working with a person, I want to be able to access all these accounts. On the other hand, an account _has one_ owner. and so I only need access to that one person. But here is the rub: in the graph, all I have is a list of edges relating people and accounts. I don't carry with me have the context of which direction I'm traversing those edges.

## Approach

Use SDL(schema definition language)! It turns out that GraphQL has a very concise format for expressing basic types and how they are related. Not only that, but its parser is easy to use from JavaScript.

We can use a schema expressed as SDL to automatically create a graph generator, and then when it comes time to navigate the generated graph, we can leverage that schema information to provide context around it. Let's see how the above problems are resolved with SDL.

### ✅ Relationship distributions are categorized and easy, simple data gen is simple.

Because we context around what time of relatonship is being generated, we can impose very helpful constraints on the distribution to use:
**one, required**: -> `constant(1)`

``` graphql
type Account {
  owner: Person!
}
```

* **one, nullable**: -> `weighted()`

A nullable relationship will be present 70% of the time by default (weighted([[0, .3], [1, 0.7]])

``` graphql
type Account {
  beneficiary: Person
}
```

but you can specify a different probability if you want, but even then it is clear what you're talking about

``` graphql
type Account {
  beneficiary: Person @has(chance: .25) #25% of accounts have beneficiaries
}
```

* **many**: -> `normal(), et, al...`

By the same token, if we detect a list relationship, we can use a default normal distribution:

``` graphql
type Person {
  accounts: [Account]
}
```

And if we want to override the default:

``` graphql
type Person {
  accounts: [Account] @normal(mean: 5, max: 100, standardDeviation: 3)
}
```

Simple data generation of field is simple:. Rather than create a distribution for the data that has to individually create all the data for that vertex, we can attach a generator per/field

``` graphql
type Person {
  name: String! @gen(with: "@faker/name.findName")
}
```

### ✅ Relationship presets have the same name as the relationship reference.

Because we have the schema information, and we can resolve type references into relationships, we can just use whatever was specified by each side of the reference

Before:
``` javascript
createVertex(graph, "Person", {
  name: "Bob",
  "Person.manager": {
    name: "Judith"
  }
});
```

After:

``` javascript
type Person {
  name: String!
  manager: Person!
  reports: [Person]
}
```

``` javascript
create("Person", {
  name: "Bob",
  manager: {
    name: "Judith",
  }
});
```

### ✅ Relationship names are unique, and managed by grapghen itself and so do not rely on fragile programmer-created conventions

Whereas before we had to keep track in our head the names of relationships so that they could be used in presets, and in graph dereferences, we don't need this anymore because we can generate them from the schema.

* unidirectional relationships get names like `Account.owner->[Person|Organization]`
* bidirectional relationships get names like `Account.owner->[Person.accounts|Organization.accounts]`

But this is really just a nice-to-know, because when we use a schema based graphgen, we never need to care about what the names of the edges are under the covers.

### ✅ Relationships are traversed automatically.

Because we have schema information, we can automatically resolve types, and we don't have to do arcane dereferencing. Rather than having the `create()` function return a raw vertex, we have it return a vertex wrapped by an object that has all of its data, but also has properties to dereference its relationships:


Before:

``` javascript
let person: Vertex = createVertex("Person");
let edges = graph.from[person.id]
assert(edges);
let [managerEdge] = edges.filter(edge => edge.type === "Person.manager");
assert(managerEdge);
let manager = graph.vertices[manageredge.to];
assert(manager);
```

After:

``` javascript
let person = create("Person");
person.manager;
```

### ✅Computed properties mean fields are derived the same no matter which order in which they were generated.

Instead of trying to track which direction a vertex generation came from, so that it can relate to its source( think, I want a github account login to have the email address of its owner). We can just derive that data directly with a function.


### ✅Has the capability to pass an API into the graphgen so that created objects are strongly typed.

If you pass a type to `createGraphGen()` it will assume that it is the key/vaule API of the vertex types. It can

``` typescript
interface API {
  Person: {
    name: string;
  }
let { create } createGraph<API>();

// @ts-expect-error
create("Person").wut
```

### Possible Drawbacks or Risks

* Cod-gen is needed to get strong typing because we have to have the interface before. I'm not a huge fan of codegen as it couples your runtime to a linux builstep and is disqualifying for edge-native apps.

* Language/Completion tools do not work seamlessly across the SDL/TS divide. For example, how do you discover which methods are available to you in the `@gen()` directive, and what their potential arguments are.

### TODOs and Open Questions

- [x] Presets will not work on circular relationships, this is a problem with graphgen and not just this syntax.
- [x] Better error handling of syntax errors in SDL.
- [ ] use Zod or some other declarative parse to lift the raw SDL into the intermediate format.
- [x] Expand on the bullet points in this PR both in explanation and solution, also search the changes for other key problems solved by this approach.
- [ ] Some level of README / Guide explaining the new Syntax

cowboyd committed:
basic configurable field-level generation

cowboyd committed:
Extract high-level field model from SDL

cowboyd committed:
analyze relationships

cowboyd committed:
create edges and relationships

cowboyd committed:
automagically resolve references

cowboyd committed:
get node refs working

cowboyd committed:
Get many to many going

cowboyd committed:
Add ability to add own custom size distribution

cowboyd committed:
polymorphic relationships

cowboyd committed:
Add typings API

cowboyd committed:
Make build work for ESM

cowboyd committed:
Make the default generation less hokey

cowboyd committed:
Add dispatch for computed and generated values

cowboyd committed:
Add ability to compute fields

cowboyd committed:
Support relationships that do _not_ have inverses

cowboyd committed:
formatting

dagda1 committed:
replace NPM_VERSION with version argument

dagda1 committed:
any ComputeMap

cowboyd committed:
Pass arguments to @gen directive

cowboyd committed:
Rename `FieldGen` -> `Generate` to contrast `Compute`

cowboyd committed:
Allow passing deeply nested presets with automatic translation

cowboyd committed:
fix test name

cowboyd committed:
Remove obsolete test case

cowboyd committed:
Add typechecking to the presets

dagda1 committed:
add min to size directive

dagda1 committed:
make ComputeMap a mapped type of correct string keys and function argument type

dagda1 committed:
add all function to return a generated type

dagda1 committed:
add createMany

dagda1 committed:
don't call computed for preset properties

dagda1 committed:
make ComputeMap optional

dagda1 committed:
use vertex.data instead of pretext to filter computed

dagda1 committed:
use vertex.data keys to filter computed

cowboyd committed:
Add `__typename` properties to `Node` API.

When simulating the backstage catalog API, you need to have the `kind`
parameter in the entity body which is the same as the type of the
graphql gen.

graphql:

```graphql
type Component {
 name: String!
 description: String!
}

backstage entity:

```json
{
  "kind": "Component",
  "metadata": {
    "name": "artists-lookup",
    "namespace": "default"
  },
  spec: {
    "description": "Artists Lookup Service"
  }
}
```

But currently the string representing the graphql type name is not
available to use.

This adds a `__typename` property which is prefixed to indicate that
it is mostly for metaprogramming. This property is deliberately not
enumerable so that it doesn't make debug output unsightly, and also so
that you can iterate over just the "concrete" properties without
making exceptions for the `__typename`

cowboyd committed:
Add `__typename` properties to `Node` API.

When simulating the backstage catalog API, you need to have the `kind`
parameter in the entity body which is the same as the type of the
graphql gen.

graphql:

```graphql
type Component {
 name: String!
 description: String!
}

backstage entity:

```json
{
  "kind": "Component",
  "metadata": {
    "name": "artists-lookup",
    "namespace": "default"
  },
  spec: {
    "description": "Artists Lookup Service"
  }
}
```

But currently the string representing the graphql type name is not
available to use.

This adds a `__typename` property which is prefixed to indicate that
it is mostly for metaprogramming. This property is deliberately not
enumerable so that it doesn't make debug output unsightly, and also so
that you can iterate over just the "concrete" properties without
making exceptions for the `__typename`

dagda1 committed:
pass ListValue as args to @gen

cowboyd committed:
read null values from graphql args.

For whatever reason, sometimes you need to pass null arguments to
faker. E.g.

```graphql
@gen(with: "@faker/image.business", args: [null,null, true])
```

This just maps GraphQL `NullValue` to JavaScript `null`

cowboyd committed:
resolve test conflicts, deno format according v1

dekoding reviewed (approved):
This looks incredible!

cowboyd committed:
Allow presets for "many" fields

We are transforming only subhashes which implies that you can only
have "one" relationships. However, you can also have "many"
relationships in which the preset will be array data.

This maps any arrays to the transform instead of treating them like
records.

cowboyd committed:
Allow array presets

cowboyd committed:
Give more info on GraphQL errors

cowboyd committed:
Formatting

cowboyd committed:
Don't include the test file from the NPM package

cowboyd committed:
🚨"fix" linting errors
